lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -366,6 +364,134 @@ static LogicalResult verifyCircuitOp(CircuitOp circuit) {
         return failure();
       }
     }
+    return success();
+  };
+
+  InnerRefList instanceSyms;
+  InnerRefList leafInnerSyms;
+  SmallVector<NonLocalAnchor> nlaList;
+
+  for (auto &op : *circuit.getBody()) {
+    // Record all the NLAs.
+    if (auto nla = dyn_cast<NonLocalAnchor>(op))
+      nlaList.push_back(nla);
+    else if (auto mod = dyn_cast<FModuleOp>(op)) {
+      auto modName = mod.getNameAttr();
+      mod.walk([&](Operation *op) {
+        if (isa<InstanceOp>(op))
+          instanceSyms.insert(op, modName);
+        else
+          leafInnerSyms.insert(op, modName);
+      });
+    }
+    if (auto mod = dyn_cast<FModuleLike>(op))
+      leafInnerSyms.insert(mod);
+    // Verify external modules.
+    if (auto extModule = dyn_cast<FExtModuleOp>(op))
+      if (verifyExtModule(extModule).failed())
+        return failure();
+  }
```

Comment:
`auto mod = dyn_cast<FModuleLike>(op)` ends up being a hashtable lookup to check if the op implements the interface, I don't think its worth it just to avoid duplicating the single line `leafInnerSyms.insert(mod);`, but this is up to you

Reply:
Apparently the `InterfaceMap::lookup` function is actually a binary search on a sorted list: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Support/InterfaceSupport.h#L264

Reply:
Given that this NLA verifier is expensive, we should try to avoid any obvious redundant work. 
Fixed it.